[Merged by Bors] - feat(DedekindDomain/Ideal): `relNorm` of a prime ideal

xroblot opened the pull request:
The relative norm of the maximal ideal `P` above the maximal ideal `p` is equal to `p ^ p.inertiaDeg P`.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

- [x] depends on: #27105
- [x] depends on: #27157
- [x] depends on: #28169
- [x] depends on: #28738 
- [x] depends on: #28724
- [x] depends on: #29068
- [x] depends on: #29069
- [x] depends on: #29907
- [x] depends on:  #30584

xroblot committed:
clean

xroblot committed:
Merge remote-tracking branch 'upstream/master' into relnorm_prime

xroblot committed:
whitespace

xroblot commented on Mathlib/RingTheory/Ideal/Norm/RelNorm.lean:
```diff
@@ -372,6 +375,151 @@ theorem relNorm_algebraMap' {R'} [CommRing R'] (I : Ideal R') [Algebra R' R]
+    exists_maximal_ideal_liesOver_of_isIntegral P
+  have : IsGalois L (FractionRing T) := IsGalois.tower_top_of_isGalois K L (FractionRing T)
+  have : Q.LiesOver p := LiesOver.trans Q P p
+  have h := relNorm_eq_pow_of_isPrime_isGalois Q p
```
The proof is done in two steps: the first step is to prove the results for an extension of Dedekind domains `R ⊆ S` such that `Frac S / Frac R` is Galois and then, in a second step, to prove the general result by constructing a Dedekind domain `T` such that `R ⊆ S ⊆ T` and `Frac T / Frac R` is Galois. This is a bit tricky and, in fact,
it is necessary to slightly increase the `maxHeartbeats` for the proof to complete.

Anyway, my question is the following, is this construction of embedding an extension `R ⊆ S` of domain into a larger domain whose fraction field is a Galois extension of `Frac R` something that should be done in a separate file since it might be useful in some other cases

riccardobrasca commented on Mathlib/RingTheory/UniqueFactorizationDomain/NormalizedFactors.lean:
```diff
@@ -214,6 +214,9 @@ theorem normalizedFactors_of_irreducible_pow {p : α} (hp : Irreducible p) (k :
 theorem zero_notMem_normalizedFactors (x : α) : (0 : α) ∉ normalizedFactors x := fun h =>
   Prime.ne_zero (prime_of_normalized_factor _ h) rfl
 
+theorem ne_zero_of_mem_normalizedFactors {x a : α} (hx : x ∈ normalizedFactors a) : x ≠ 0 :=
```
If you need both results no problem, but maybe we can reformulate something so that everything use the same formulation.

xroblot committed:
Merge remote-tracking branch 'upstream/master' into relnorm_prime

xroblot committed:
wip

riccardobrasca committed:
feat: define positive semidefinite sesquilinear maps (#28162)
Following what is done for `IsSymm`, we duplicate the definition for bilinear forms.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

riccardobrasca committed:
fix

riccardobrasca committed:
remove assumptions

riccardobrasca committed:
remove also this

riccardobrasca committed:
here it is needed

riccardobrasca committed:
unused

xroblot committed:
Merge remote-tracking branch 'riccardobrasca/RB/norm_integral' into relnorm_prime

xroblot committed:
wip

xroblot committed:
review

xroblot committed:
Merge remote-tracking branch 'upstream/master' into relnorm_prime

xroblot committed:
1st commit

xroblot committed:
update mathlib.lean

xroblot committed:
Merge remote-tracking branch 'upstream/master' into relnorm_prime

xroblot committed:
merge with #30584

xroblot committed:
docstring

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

xroblot committed:
scoped to local

xroblot committed:
Merge remote-tracking branch 'upstream/master' into ring_normalclosure

xroblot committed:
Merge remote-tracking branch 'origin/ring_normalclosure' into relnorm_prime

xroblot committed:
lint

xroblot committed:
Merge remote-tracking branch 'upstream/master' into relnorm_prime

riccardobrasca reviewed (commented):
Thanks!

bors d+

xroblot commented on Mathlib/RingTheory/Ideal/Norm/RelNorm.lean:
```diff
@@ -35,12 +39,12 @@ namespace Ideal
 variable [IsIntegrallyClosed R] [IsIntegrallyClosed S] [Algebra R S] [Module.Finite R S]
 variable [NoZeroSMulDivisors R S]
 
+attribute [local instance] FractionRing.liftAlgebra
```
Yes, it is still needed at several places.
